### PR TITLE
Vision port

### PR DESCRIPTION
**Pull request by loqoman**
#### How does a target go from the Rasberry pi to `GoalTrack`?
* `generateUpdate()` in `VisionServer` pulls data from `SmartDashboard` and stores it in a `TargetInfo`.
* The target is passed to `GoalTracker` through `addVisionUpdate`.
* Inside `GoalTracker`, the target is turned into a `Pose2d`.
* The conversion to `Pose2d` is where the camera offsets are implemented.
* The `Pose2d` is passed to update, which is the final step in the life of a target.
* Datastructures inside `GoalTracker` maintain lists of targets.

#### What is a `TargetInfo`
* A `TargetInfo` is a melding of 254's `VisionMessage` and `TargetInfo`. Because our implementation of Vision requires much less processing robot-side, they were turned into the same object.
* It has an x, y, z, and timestamp.
* The timestamp of the target is the `clock` value posted by the RasberryPi at the time of capture.
* The y and z values in the `TargetInfo` are equal to x and y in RasberryPi cords. I have chosen to not modify the variable names in `TargetInfo`, because they represent the change from a 2D plane (Pi) to a 3D plane (Robot)
* The `TargetInfo` data structure is subject to be renamed, because it does NOT serve the same purpose in 254's codebase.

#### What is `GoalTracker`?
* `GoalTracker` is the mirror of `GoalTrack`
* Both classes are responsible for the tracking of all vision targets discovered through the match
* I don't 100% understand how `GoalTracker` and `GoalTrack` 'weights' particular targets, and how the tracking system works

#### How are vision updates generated?
* Vision updates are generated in VisionServer (Registered as a loop)
* The function `generateUpdate()` pulls the x,y, status, and timestamp posted by the Pi. The x, y, and timestamp is pushed into a `TargetInfo`
* Currently, `generateUpdate()` only returns a non-null TargetInfo if the following conditions are true:
_x is not equal to 0
y is not equal to 0
timestamp is not equal to 0_
* The aforementioned logic **IS VERY SHADY**. Subject to change.

#### Is there compensation for a potential camera offset?
* Yes!
* All the geometry and trig is done in `GoalTracker`, which is pulling 6 main constants about the cameras position in relation to the robot.
* It is very important that the constants are updated properly once a robot model becomes clear, so the target tracking system works properly.

#### What are the new constants?
* There are 6 new constants, one for pitch, yaw, and roll, and 3 for x, y, and z of the camera in relation to the robot chassis.
* The 3 responsible for pitch, yaw and roll I understand, I don't understand the geometry behind the x, y, and z values.
* `kCameraFrameRate` is used in `GoalTrack` to weight a specific observation. (I.E: An observation measured at a low frame rate is likely to be incorrect)
* Two new constants were added to control goal weighting.
* There are some misc. constants for 2017's Steamworks. These will change for the 2019 game. They are used in `GoalTrack `to determine intersection with the Boiler (Or in our case, any objective).
* As mentioned above, it is very important to update and maintain the new constants.

#### Why is `VisionServer` a loop?
* 254 wrote `VisionServer` as a `CrashTrackingRunnable`, due to their method of communication.
* 254's communication relied on a JSON reading generated by the phone, and was received over ADB. Because of this, their code would often block waiting for input.
* Blocking code as a loop is VERY BAD. Each of the registered loops(Subsystems, Drive, Controls(?) ) is called every iteration of `looper`, and if one of them stops to run a large chunk of code waiting for input (I.E A JSON reading), then all the other loops are delayed until the blocking code is finished.
* However, our vision solution is not nearly as complicated, and requires no blocking code. Thus, a loop!

### CODE REVIEW

**Comment by dbadb** [comment]
I'm a little confused by the diffs here.  They seem to overlap with many changes already present in the mainline.  Perhaps @pietroglyph can clarify?

File: src/main/java/com/spartronics4915/frc2019/Robot.java [hunk 1]
```diff
@@ -20,8 +21,7 @@
 import java.util.jar.Attributes;
 import java.util.jar.Manifest;
 
-public class Robot extends IterativeRobot
-{
+public class Robot extends IterativeRobot {
```
Comment (dbadb) [inline src/main/java/com/spartronics4915/frc2019/Robot.java]: oops - our code formatting standard requires trailing braces

Comment (dbadb) [reply src/main/java/com/spartronics4915/frc2019/Robot.java]: many of the "noise" in this checkin appears to be due to the formatting changes.
I wonder if you can "fix" your checkin to eliminate that noise.  In other words,
we need to stick to one code formatting convention to make code reviews easier.

Comment (pietroglyph) [reply src/main/java/com/spartronics4915/frc2019/Robot.java]: @loqoman I think that's the first change we need. You can find the formatter [here](https://github.com/Spartronics4915/Spartronics-254Base2018/blob/master/eclipse-formatter.xml). You can edit VSCode's `java.format.settings.url` setting to point to that file on your disk. Then just apply the formatting to the files you've changed.

Comment (loqoman) [reply src/main/java/com/spartronics4915/frc2019/Robot.java]: I agree...Some of these diffs I don't remember even touching...

**Comment by loqoman** [comment]
I ran `format document` in VScode on Robot, all the new vision files, and Constants

File: src/main/java/com/spartronics4915/frc2019/vision/GoalTracker.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+{
+
+    /**
+     * Track reports contain all of the relevant information about a given goal
+     * track.
+     */
+
+    // TODO: Find a better home for these varibles
+
+    private Rotation2d camera_pitch_correction_;
```
Comment (dbadb) [inline src/main/java/com/spartronics4915/frc2019/vision/GoalTracker.java:36]: convert these to our mMemberVar rep?

File: src/main/java/com/spartronics4915/frc2019/vision/VisionServer.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    private boolean running_;
+
+    static VisionServer instance_ = null;
+    // RobotState robot_state_ = RobotState.getInstance();
+    // Goaltracker
+
+    String defaultValue = "null";
+
+    TargetInfo update_;
```
Comment (dbadb) [inline src/main/java/com/spartronics4915/frc2019/vision/VisionServer.java:48]: mMemberVar formatting?

File: src/main/java/com/spartronics4915/frc2019/vision/GoalTrack.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+     * @return True if the track was updated
+     */
+    public boolean tryUpdate(double timestamp, Translation2d new_observation)
+    {
+        if (!isAlive())
+        {
+            return false;
+        }
+        double distance = mSmoothedPosition.inverse().translateBy(new_observation).norm();
+        // XXX: What does kMaxTrackerDistance mean?
```
Comment (pietroglyph) [inline src/main/java/com/spartronics4915/frc2019/vision/GoalTrack.java:63]: I belive that `kMaxTrackerDistance` is the distance between the last known position of this `GoalTrack`, and the new observation.

File: src/main/java/com/spartronics4915/frc2019/vision/GoalTracker.java (hunk 2 above)
Comment (pietroglyph) [reply src/main/java/com/spartronics4915/frc2019/vision/GoalTracker.java:36]: @loqoman Btw, `F2` renames variables in VSCode, if you didn't know that already.

**Pull request closed without merging**